### PR TITLE
New analytics services for Solar PV Advice

### PR DESCRIPTION
**Pull request by iant**

### CODE REVIEW

**Comment by iant** [comment]
```
Calculated meter aggregation for |Acme Academy                       | in |0.413| seconds
=============================================================AlertSolarPVBenefitEstimator=============================================================
{
                                                                       "AlertSolarPVBenefitEstimator:optimum_kwp" => 52.5,
                                                             "AlertSolarPVBenefitEstimator:optimum_payback_years" => 5.682322708769174,
                                                   "AlertSolarPVBenefitEstimator:optimum_mains_reduction_percent" => 0.10478762755164217,
                                                          "AlertSolarPVBenefitEstimator:one_year_saving_£current" => 7182.209673170789,
                                                                         "AlertSolarPVBenefitEstimator:relevance" => :relevant,
                                                                     "AlertSolarPVBenefitEstimator:analysis_date" => #<Date: 2020-12-31 ((2459215j,0s,0n),+0s,2299161j)>,
                                                                            "AlertSolarPVBenefitEstimator:status" => nil,
                                                                            "AlertSolarPVBenefitEstimator:rating" => 5.0,
                                                                              "AlertSolarPVBenefitEstimator:term" => nil,
                                                                      "AlertSolarPVBenefitEstimator:bookmark_url" => nil,
                                                                      "AlertSolarPVBenefitEstimator:max_asofdate" => #<Date: 2022-07-13 ((2459774j,0s,0n),+0s,2299161j)>,
                                                                            "AlertSolarPVBenefitEstimator:pupils" => 961,
                                                                        "AlertSolarPVBenefitEstimator:floor_area" => 5932.0,
                                                                       "AlertSolarPVBenefitEstimator:school_type" => :secondary,
                                                                       "AlertSolarPVBenefitEstimator:school_name" => "Acme Academy",
                                                            "AlertSolarPVBenefitEstimator:school_activation_date" => nil,
                                                              "AlertSolarPVBenefitEstimator:school_creation_date" => #<Date: 2020-10-08 ((2459131j,0s,0n),+0s,2299161j)>,
                                                                               "AlertSolarPVBenefitEstimator:urn" => 136770,
                                                                 "AlertSolarPVBenefitEstimator:one_year_saving_£" => 7182.209673170789..7182.209673170789,
                                                             "AlertSolarPVBenefitEstimator:one_year_saving_£_low" => 7182.209673170789,
                                                            "AlertSolarPVBenefitEstimator:one_year_saving_£_high" => 7182.209673170789,
                                                               "AlertSolarPVBenefitEstimator:one_year_saving_co2" => 8884.491460063697,
                                                               "AlertSolarPVBenefitEstimator:ten_year_saving_co2" => 88844.91460063697,
                                                         "AlertSolarPVBenefitEstimator:average_one_year_saving_£" => 7182.209673170789,
                                                         "AlertSolarPVBenefitEstimator:average_ten_year_saving_£" => 71822.09673170789,
                                                                 "AlertSolarPVBenefitEstimator:ten_year_saving_£" => 71822.09673170789..71822.09673170789,
                                                             "AlertSolarPVBenefitEstimator:ten_year_saving_£_low" => 71822.09673170789,
                                                            "AlertSolarPVBenefitEstimator:ten_year_saving_£_high" => 71822.09673170789,
                                                                     "AlertSolarPVBenefitEstimator:payback_years" => nil,
                                                             "AlertSolarPVBenefitEstimator:average_payback_years" => 5.682322708769174,
                                                                      "AlertSolarPVBenefitEstimator:capital_cost" => 40811.633125..40811.633125,
                                                                  "AlertSolarPVBenefitEstimator:capital_cost_low" => 40811.633125,
                                                                 "AlertSolarPVBenefitEstimator:capital_cost_high" => 40811.633125,
                                                              "AlertSolarPVBenefitEstimator:average_capital_cost" => 40811.633125,
                                                                         "AlertSolarPVBenefitEstimator:timescale" => "year",
                                                            "AlertSolarPVBenefitEstimator:time_of_year_relevance" => 5.0,
                                             "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Panels" => 3,
                                          "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Area (m2)" => 4,
       "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Annual self consumed solar electricity (kWh)" => 893.9545973935678,
            "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Annual exported solar electricity (kWh)" => 0.0,
                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Annual output from panels (kWh)" => 893.954597393567,
                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Reduction in mains consumption" => 0.002068165948887468,
                                      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Annual saving" => 140.1736542641811,
                                "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Annual saving (CO2)" => 169.22840876311736,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Estimated cost" => 2392.9653,
                                      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:1:Payback years" => 17.07143409053209,
                                             "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Panels" => 7,
                                          "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Area (m2)" => 10,
       "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Annual self consumed solar electricity (kWh)" => 1787.9091947871332,
            "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Annual exported solar electricity (kWh)" => 0.0,
                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Annual output from panels (kWh)" => 1787.909194787134,
                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Reduction in mains consumption" => 0.004136331897774667,
                                      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Annual saving" => 280.34730852871144,
                                "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Annual saving (CO2)" => 338.4568175262347,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Estimated cost" => 3184.1412,
                                      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:2:Payback years" => 11.357844727351466,
                                             "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Panels" => 13,
                                          "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Area (m2)" => 19,
       "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Annual self consumed solar electricity (kWh)" => 3575.8183895742663,
            "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Annual exported solar electricity (kWh)" => 0.0,
                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Annual output from panels (kWh)" => 3575.818389574268,
                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Reduction in mains consumption" => 0.008272663795547853,
                                      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Annual saving" => 560.6946170564115,
                                "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Annual saving (CO2)" => 676.9136350524694,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Estimated cost" => 4761.1248,
                                      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:4:Payback years" => 8.491475850072202,
                                             "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Panels" => 27,
                                          "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Area (m2)" => 39,
       "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Annual self consumed solar electricity (kWh)" => 7151.636779148539,
            "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Annual exported solar electricity (kWh)" => 0.0,
                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Annual output from panels (kWh)" => 7151.636779148536,
                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Reduction in mains consumption" => 0.016545327591100552,
                                      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Annual saving" => 1121.3892341130122,
                                "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Annual saving (CO2)" => 1353.8272701049389,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Estimated cost" => 7893.6192,
                                      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:8:Payback years" => 7.0391430199913,
                                            "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Panels" => 53,
                                         "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Area (m2)" => 76,
      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Annual self consumed solar electricity (kWh)" => 14303.273558297074,
           "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Annual exported solar electricity (kWh)" => 0.0,
                   "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Annual output from panels (kWh)" => 14303.273558297073,
                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Reduction in mains consumption" => 0.03309065518220407,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Annual saving" => 2242.778468225384,
                               "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Annual saving (CO2)" => 2707.6545402098777,
                                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Estimated cost" => 14072.7168,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:16:Payback years" => 6.274679822093685,
                                            "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Panels" => 107,
                                         "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Area (m2)" => 154,
      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Annual self consumed solar electricity (kWh)" => 28554.856062377454,
           "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Annual exported solar electricity (kWh)" => 51.691054216699065,
                   "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Annual output from panels (kWh)" => 28606.547116594145,
                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Reduction in mains consumption" => 0.06606172299553545,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Annual saving" => 4479.975043753114,
                               "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Annual saving (CO2)" => 5415.309080419755,
                                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Estimated cost" => 26087.3472,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:32:Payback years" => 5.823101009541616,
                                          "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Panels" => 175,
                                       "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Area (m2)" => 252,
    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Annual self consumed solar electricity (kWh)" => 45293.93854982251,
         "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Annual exported solar electricity (kWh)" => 1638.67781333994,
                 "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Annual output from panels (kWh)" => 46932.61636316246,
                  "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Reduction in mains consumption" => 0.10478762755164217,
                                   "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Annual saving" => 7182.209673170789,
                             "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Annual saving (CO2)" => 8884.491460063697,
                                  "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Estimated cost" => 40811.633125,
                                   "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:52.5:Payback years" => 5.682322708769174,
                                            "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Panels" => 213,
                                         "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Area (m2)" => 307,
      "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Annual self consumed solar electricity (kWh)" => 53186.555323147804,
           "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Annual exported solar electricity (kWh)" => 4026.538910040475,
                   "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Annual output from panels (kWh)" => 57213.09423318829,
                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Reduction in mains consumption" => 0.12304721400692871,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Annual saving" => 8535.883922746882,
                               "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Annual saving (CO2)" => 10830.61816083951,
                                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Estimated cost" => 48742.3488,
                                     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:64:Payback years" => 5.710287211159089,
                                           "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Panels" => 427,
                                        "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Area (m2)" => 615,
     "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Annual self consumed solar electricity (kWh)" => 85201.53193961504,
          "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Annual exported solar electricity (kWh)" => 29224.65652676189,
                  "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Annual output from panels (kWh)" => 114426.18846637658,
                   "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Reduction in mains consumption" => 0.1971139335983544,
                                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Annual saving" => 14787.612399356698,
                              "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Annual saving (CO2)" => 21661.23632167902,
                                   "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Estimated cost" => 88555.3152,
                                    "AlertSolarPVBenefitEstimator:AlertSolarPVBenefitEstimator:128:Payback years" => 5.9884796009295185
}
```

File: app/services/solar_photovoltaics/benefit_estimator_service.rb [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    def initialize(school:, asof_date: Date.today)
+      @school = school
+      raise if @school.solar_pv_panels?
+      raise unless enough_data?
+
+      @asof_date = asof_date
+    end
+
+    def calculate_benefits!
```
Comment (ldodds) [inline app/services/solar_photovoltaics/benefit_estimator_service.rb:15]: Why use a ! here? Generally I though `!` naming was to indicate that the state of the object being called will change.

File: app/services/solar_photovoltaics/existing_benefits_service.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module SolarPhotovoltaics
+  class ExistingBenefitsService
+    def initialize(meter_collection:)
+      @meter_collection = meter_collection
+    end
+
+    def calculate
```
Comment (iant) [inline app/services/solar_photovoltaics/existing_benefits_service.rb:9]: shall we call this `create_model` instead of `calculate`?

**Review by ldodds — APPROVED** [review]

File: app/services/solar_photovoltaics/benefit_estimator_service.rb (hunk 1 above)
Comment (iant) [reply app/services/solar_photovoltaics/benefit_estimator_service.rb]: :+1: I've update to just return an object here

**Comment by iant** [comment]
hi @ldodds could you give this another review when you get a mo please

**Review by ldodds — APPROVED** [review]

**Pull request merged**